k8s_test_harness/plugin.py

```diff
@@ -26,37 +26,47 @@ def _harness_clean(h: harness.Harness):
         h.cleanup()
 
 
-@pytest.fixture(scope="module")
-def h() -> harness.Harness:
-    LOG.debug("Create harness for %s", config.SUBSTRATE)
-    if config.SUBSTRATE == "local":
-        h = harness.LocalHarness()
-    elif config.SUBSTRATE == "lxd":
-        h = harness.LXDHarness()
-    elif config.SUBSTRATE == "multipass":
-        h = harness.MultipassHarness()
-    elif config.SUBSTRATE == "juju":
-        h = harness.JujuHarness()
-    else:
+def _get_harness() -> harness.Harness:
+    LOG.debug("Initializing harness for %s", config.SUBSTRATE)
+    harness_cls_map = {
+        "local": harness.LocalHarness,
+        "lxd": harness.LXDHarness,
+        "multipass": harness.MultipassHarness,
+        "juju": harness.JujuHarness,
+    }
+    harness_cls = harness_cls_map.get(config.SUBSTRATE)
+    if not harness_cls:
         raise harness.HarnessError(
             "TEST_SUBSTRATE must be one of: local, lxd, multipass, juju"
         )
 
+    return harness_cls()
+
+
+@pytest.fixture(scope="module")
+def h_module() -> harness.Harness:
+    h = _get_harness()
+
     yield h
 
     _harness_clean(h)
 
 
-@pytest.fixture(scope="module")
-def module_instance(
-    h: harness.Harness, tmp_path_factory: pytest.TempPathFactory, request
-) -> Generator[harness.Instance, None, None]:
+@pytest.fixture(scope="function")
+def h_function() -> harness.Harness:
```

Comment:
You don't need a harness for function and module scope. Just do something like here:
https://github.com/canonical/k8s-snap/blob/main/tests/integration/tests/conftest.py#L73

Reply:
Ohh, I see. The harness might as well be session scoped, it's the instance scope that matters. Thanks!